### PR TITLE
0x6c Chapter Restructure

### PR DESCRIPTION
**Pull request by cpholguera**
Done in this PR:

- Chapter 0x6c restructured according to issue #970. 
- Contents are the SAME, just have the new structure and some had to be adapted or made consistent with the structure so some lines might be new.
- Frida example in "##### Method Hooking" was reviewed and corrected (it wasn't working properly).
- Some other chapters also have minor changes due to consistency of images location.
- Several fixes to comply with the style guide.
- "runtime" added to the style guide list of words.

Some content was moved from 6b to 6c:
- Ghidra.
- Static and Dynamic analysis sections.
- Method tracing with Frida and other Frida information.

Please note that there are several missing sections according to the new structure. They will be added as new issues including description and where they have to be located in the new outline **once this PR is merged** to avoid confusion. However, **they're not added inline to the chapter as --TODOs** (as we did for the OSS19) so that we don't disturb the release process and keep adding new sections for newer releases.

This PR covers issue #970.

The following picture should help when reviewing this PR (left = current master; right = restructured version):

Screenshot 2019-07-17 at 22 27 34

Final Outline after completing all new tickets:

```
## Tampering and Reverse Engineering on iOS
### Reverse Engineering
#### Tooling
##### Building a Reverse Engineering Environment for Free
##### Commercial Tools
###### IDA Pro
###### Hopper
#### Disassembling and Decompiling
##### Disassembling Objective-C and Swift Code
##### Disassembling Native Code
###### radare2
###### IDA Pro
### Static Analysis
#### Manual (Reversed) Code Review
##### Reviewing Disassembled Objective-C and Swift Code
##### Reviewing Disassembled Native Code
###### radare2
###### IDA Pro
#### Basic Information Gathering
##### Strings
##### Call Diagrams and Cross References
##### API Usage
##### Check Secure Connections
#### Automated Static Analysis
### Dynamic Analysis
#### Dynamic Analysis on Non-Jailbroken Devices
##### Automated Repackaging with Objection
##### Manual Repackaging
##### Getting a Developer Provisioning Profile and Certificate
##### Other Preparations
#### Basic Information Gathering
##### Opened Files
##### Opened Connections
##### Loaded Native Libraries
##### Sandbox Inspection
#### Debugging
##### Debugging Release Apps
##### Debugging with lldb
##### Debugging with gdb
##### Debugging with an IDE
##### Debugging Native Code
#### Tracing
##### Execution Tracing
##### Method Tracing
##### Native Libraries Tracing
#### Emulation-based Analysis
##### iOS Simulator
##### Corellium ???
#### Binary Analysis
##### Symbolic Execution
### Tampering and Runtime Instrumentation
#### Patching, Repackaging, and Re-Signing
##### Patching Example: Installing Frida Gadget
##### Patching Example: Making an App Debuggable
##### Repackaging and Re-Signing
##### Patching React Native applications
#### Dynamic Instrumentation
##### Tooling
###### Frida
###### Cycript
##### Information Gathering
###### Getting Loaded Libraries
###### Getting Loaded Classes and their Methods
###### Getting Runtime Dependencies
##### Method Hooking
###### Frida
###### Cycript
##### Process Exploration
###### Memory Maps and Inspection
###### In-Memory Search
###### Memory Dump
###### Runtime Reverse Engineering
### References
#### Tools
```

### CODE REVIEW

**Comment by OMTGreleaser** [comment]
Results for commit 46a4eb5fe126304e9cf47d897c2fcb8976300f7f: Broken link result: [!] Error(s) found by the Links validator: 3 pages have dead links! Verbose output in /link-check-result.out .
 Please run the Apply_Link_Check.sh in the tools folder to see wherhe the Link issues are. 
 Markdown result errorlength:        1 ../linter-result.out .
 Please run the Apply_Linter_check.sh in the tools folder to see where the Lint issues are.

**Comment by OMTGreleaser** [comment]
Results for commit 718e7466f11c3353301e0d29bb716014b9230de5: Broken link result: [!] Error(s) found by the Links validator: 3 pages have dead links! Verbose output in /link-check-result.out .
 Please run the Apply_Link_Check.sh in the tools folder to see wherhe the Link issues are. 
 Markdown result errorlength:        1 ../linter-result.out .
 Please run the Apply_Linter_check.sh in the tools folder to see where the Lint issues are.

File: Document/0x06c-Reverse-Engineering-and-Tampering.md [hunk 1]
```diff
@@ -426,17 +224,114 @@ zsh: # . ~/.zshrc
...
+
+##### Debugging with lldb
+
+iOS ships with the console app debugserver, which allows remote debugging via gdb or lldb. By default, however, debugserver can't be used to attach to arbitrary processes (it is usually used only for debugging self-developed apps deployed with Xcode). To enable debugging of third-party apps, the `task_for_pid` entitlement must be added to the debugserver executable. An easy way to do this is to add the entitlement to the [debugserver binary shipped with Xcode](http://iphonedevwiki.net/index.php/Debugserver "Debug Server on the iPhone Dev Wiki").
+
+To obtain the executable, mount the following DMG image:
 
 ```shell
-$ curl -O https://build.frida.re/frida/ios/lib/FridaGadget.dylib
+/Applications/Xcode.app/Contents/Developer/Platforms/iPhoneOS.platform/DeviceSupport/<target-iOS-version/DeveloperDiskImage.dmg
```
Comment (TheDauntless) [inline Document/0x06c-Reverse-Engineering-and-Tampering.md]: Incomplete tag

Comment (cpholguera) [reply Document/0x06c-Reverse-Engineering-and-Tampering.md]: good catch, thanks ;)

**Comment by OMTGreleaser** [comment]
Results for commit 7be543f2b43d74134cd6c0e929548f95a7fc32f5: Broken link result: [!] Error(s) found by the Links validator: 4 pages have dead links! Verbose output in /link-check-result.out .
 Please run the Apply_Link_Check.sh in the tools folder to see wherhe the Link issues are. 
 Markdown result errorlength:        1 ../linter-result.out .
 Please run the Apply_Linter_check.sh in the tools folder to see where the Lint issues are.

**Comment by OMTGreleaser** [comment]
Results for commit 42f12e35fc87de6b8a6f7d551af8d363f990e591: Broken link result: [!] Error(s) found by the Links validator: 5 pages have dead links! Verbose output in /link-check-result.out .
 Please run the Apply_Link_Check.sh in the tools folder to see wherhe the Link issues are. 
 Markdown result errorlength:        1 ../linter-result.out .
 Please run the Apply_Linter_check.sh in the tools folder to see where the Lint issues are.

File: Document/0x06b-Basic-Security-Testing.md [hunk 2]
```diff
@@ -178,10 +178,6 @@ In order to analyze iOS apps, you should install the following tools on your hos
 
 [Frida-ios-dump](https://github.com/AloneMonkey/frida-ios-dump "frida-ios-dump") allows you to pull a decrypted IPA from a jailbroken device. Please refer to the section "Using Frida-ios-dump" for detailed instructions on how to use it.
 
-##### Ghidra
```
Comment (commjoen) [inline Document/0x06b-Basic-Security-Testing.md]: i assume this is removed because this is for reverse-engineering right?

File: Document/0x06b-Basic-Security-Testing.md [hunk 3]
```diff
@@ -1332,150 +1339,6 @@ The last step would be to set the proxy globally on your iOS device:
 
 Open Safari and go to any webpage, you should see now the traffic in Burp. Thanks @hweisheimer for the [initial idea](https://twitter.com/hweisheimer/status/1095383526885724161 "Port Forwarding via USB on iOS")!
 
-##### Using Frida
```
Comment (commjoen) [inline Document/0x06b-Basic-Security-Testing.md]: is it correct that these materials will be moved?

File: Document/0x06c-Reverse-Engineering-and-Tampering.md [hunk 4]
```diff
@@ -500,56 +393,163 @@ PID  Name
...
+  |    |    | <_UITextFieldRoundedRectBackgroundViewNeue: 0x16e92770; frame = (0 0; 304 30); opaque = NO; autoresize = W+H; userInteractionEnabled = NO; layer = <CALayer: 0x16e92990>>
+  |    | <UIButton: 0x16d901e0; frame = (8 191; 304 30); opaque = NO; autoresize = RM+BM; layer = <CALayer: 0x16d90490>>
+  |    |    | <UIButtonLabel: 0x16e72b70; frame = (133 6; 38 18); text = 'Verify'; opaque = NO; userInteractionEnabled = NO; layer = <_UILabelLayer: 0x16e974b0>>
+  |    | <_UILayoutGuide: 0x16d92a00; frame = (0 0; 0 20); hidden = YES; layer = <CALayer: 0x16e936b0>>
+  |    | <_UILayoutGuide: 0x16d92c10; frame = (0 568; 0 0); hidden = YES; layer = <CALayer: 0x16d92cb0>>`
+```
+
+##### Method Hooking
+
+###### Frida
```
Comment (commjoen) [inline Document/0x06c-Reverse-Engineering-and-Tampering.md]: this feels like a bit too deep...

File: Document/0x06c-Reverse-Engineering-and-Tampering.md [hunk 5]
```diff
@@ -584,40 +588,41 @@ frida_code = """
...
 
-Please also take a look at the [Frida JavaScript API reference](https://www.frida.re/docs/javascript-api/).
+###### Cycript
 
-### Patching React Native Applications
-
-If the [React Native](https://facebook.github.io/react-native "React Native") framework has been used for development, the main application code is in the file `Payload/[APP].app/main.jsbundle`. This file contains the JavaScript code. Most of the time, the JavaScript code in this file is minified. With the tool [JStillery](https://mindedsecurity.github.io/jstillery "JStillery"), a human-readable version of the file can be retried, which will allow code analysis. The [CLI version of JStillery](https://github.com/mindedsecurity/jstillery/ "CLI version of JStillery") and the local server are preferable to the online version because the latter discloses the source code to a third party.
+- Install the application that will be hooked.
```
Comment (commjoen) [inline Document/0x06c-Reverse-Engineering-and-Tampering.md]: it is a bit odd to have enumeration directly here...

File: Document/0x06b-Basic-Security-Testing.md [hunk 6]
```diff
@@ -1281,25 +1277,36 @@ Keychain Data: WOg1DfuH
...
+Starting device <UDID> [SUCCEEDED] with interface rvi0
+```
 
-##### Automated Static Analysis Tools
+1. Launch Wireshark and select "rvi0" as the capture interface.
+1. Filter the traffic with Capture Filters in Wireshark to display what you want to monitor (for example, all HTTP traffic sent/received via the IP address 192.168.1.1).
 
-Several automated tools for analyzing iOS apps are available; most of them are commercial tools. The free and open source tools [MobSF](https://github.com/MobSF/Mobile-Security-Framework-MobSF "Mobile Security Framework (MobSF)") and [Needle](https://github.com/mwrlabs/needle "Needle") have some static and dynamic analysis functionality. Additional tools are listed in the "Static Source Code Analysis" section of the "Testing Tools" appendix.
+```shell
+ip.addr == 192.168.1.1 && http
```
Comment (commjoen) [inline Document/0x06b-Basic-Security-Testing.md]: this is not really shell right? can we use something like txt? I think a little image might help here for first-users. After all: the filter is the bar directly visual.

File: Document/0x06c-Reverse-Engineering-and-Tampering.md [hunk 7]
```diff
@@ -1,84 +1,64 @@
...
+- swift-demangle is an Xcode tool that demangles Swift symbols. For more information run `xcrun swift-demangle -help` once installed.
+- simctl is an Xcode tool that allows you to interact with iOS simulators via the command line to e.g. manage simulators, launch apps, take screenshots or collect their logs.
 
-```shell
-iPhone:~ root# ipainstaller -b com.example.targetapp -o /tmp/example.ipa
-```
+##### Commercial Tools
 
-###### From Non-Jailbroken Devices
+Although working with a completely free setup is possible, you should consider investing in commercial tools. The main advantage of these tools is convenience: they come with a nice GUI, lots of automation, and end user support. If you earn your daily bread as a reverse engineer, they will save you a lot of time.
```
Comment (commjoen) [inline Document/0x06c-Reverse-Engineering-and-Tampering.md]: The first sentence here is a bit odd: can you rephrase this paragraph, because the overall idea of the paragraph (doing reverse engineering? then a commercial tool can help) is nice, but i am sure you don't need it if you already have enough support with the free tools available...

File: Document/0x06c-Reverse-Engineering-and-Tampering.md [hunk 8]
```diff
@@ -426,17 +222,114 @@ zsh: # . ~/.zshrc
...
 ```
 
-We'll be using standard tools that come with macOS and Xcode in addition to the tools mentioned above. Make sure you have the [Xcode command line developer tools](https://railsapps.github.io/xcode-command-line-tools.html "Xcode Command Line Tools") installed.
+#### Tracing
+
+##### Execution Tracing
+
+Intercepting Objective-C methods is a useful iOS security testing technique. For example, you may be interested in data storage operations or network requests. In the following example, we'll write a simple tracer for logging HTTP(S) requests made via iOS standard HTTP APIs. We'll also show you how to inject the tracer into the Safari web browser.
+
+In the following examples, we'll assume that you are working on a jailbroken device. If that's not the case, you first need to follow the steps outlined in the previous section to repackage the Safari app.
```
Comment (commjoen) [inline Document/0x06c-Reverse-Engineering-and-Tampering.md]: perhaps refer to the section at hand?

File: Document/0x06c-Reverse-Engineering-and-Tampering.md [hunk 9]
```diff
@@ -500,56 +393,163 @@ PID  Name
...
 (...)
-21324 ms  -[NSURLRequest initWithURL:0x106388b00 ]
-21324 ms     | -[NSURLRequest initWithURL:0x106388b00 cachePolicy:0x0 timeoutInterval:0x106388b80
 ```
 
+We will demonstrate a few more uses for Frida throughout the chapter.
+
+###### Cycript
+
+Cydia Substrate (formerly called MobileSubstrate) is the standard framework for developing runtime patches ("Cydia Substrate extensions") on iOS. It comes with Cynject, a tool that provides code injection support for C. Cycript is a scripting language developed by Jay Freeman (aka saurik). It injects a JavaScriptCore VM into the running process. Via the Cycript interactive console, users can then manipulate the process with a hybrid Objective-C++ and JavaScript syntax. Accessing and instantiating Objective-C classes inside a running process is also possible. Examples of Cycript usage are included in the iOS chapter.
```
Comment (commjoen) [inline Document/0x06c-Reverse-Engineering-and-Tampering.md]: aka sauri--> aka Saurik (Capitalize?)

File: Document/0x06c-Reverse-Engineering-and-Tampering.md [hunk 10]
```diff
@@ -500,56 +393,163 @@ PID  Name
...
+
+```
+
+Use the following command to get the application's delegate class:
+
+```shell
+cy# [UIApplication sharedApplication].delegate
+```
+
+The command `[[UIApp keyWindow] recursiveDescription].toString()` returns the view hierarchy of keyWindow. The description of every subview and sub-subview of keyWindow is shown. The indentation space reflects the relationships between views. For example, UILabel, UITextField, and UIButton are subviews of UIView.
```
Comment (commjoen) [inline Document/0x06c-Reverse-Engineering-and-Tampering.md]: keyWindow with marking? `keyWindow`

Comment (commjoen) [reply Document/0x06c-Reverse-Engineering-and-Tampering.md]: Same holds for the other classnames

File: Document/0x06c-Reverse-Engineering-and-Tampering.md [hunk 11]
```diff
@@ -575,6 +575,10 @@ frida_code = """
                 // Create an immutable ObjC string object from a JS string object.
                 var str_url = NSString.stringWithString_(myNSURL.toString());
                 NSLog(str_url);
+
+                // Print the URL to the standard console
+                console.log(str_url);
```
Comment (commjoen) [inline Document/0x06c-Reverse-Engineering-and-Tampering.md]: mayb eexplain differnece between console.log and NSLog here?

File: Document/0x06b-Basic-Security-Testing.md (hunk 2 above)
Comment (cpholguera) [reply Document/0x06b-Basic-Security-Testing.md]: yes, it was just moved to 6c :)

File: Document/0x06b-Basic-Security-Testing.md (hunk 3 above)
Comment (cpholguera) [reply Document/0x06b-Basic-Security-Testing.md]: yes, this is all about tracing and hooking which is all now in 6c in the Dynamic Analysis section(same as in 5c now)

File: Document/0x06c-Reverse-Engineering-and-Tampering.md (hunk 4 above)
Comment (cpholguera) [reply Document/0x06c-Reverse-Engineering-and-Tampering.md]: it's the same now in 5c, we can leave it to be consistent and then think how we can make both less deep?

File: Document/0x06c-Reverse-Engineering-and-Tampering.md (hunk 5 above)
Comment (cpholguera) [reply Document/0x06c-Reverse-Engineering-and-Tampering.md]: Actually I agree. That was already there but I can fix it by putting a sentence before.

**Comment by OMTGreleaser** [comment]
Results for commit b1b19262d16f7acb37f03ce512588af5934b70e9: Broken link result: [!] Error(s) found by the Links validator: 4 pages have dead links! Verbose output in /link-check-result.out .
 Please run the Apply_Link_Check.sh in the tools folder to see wherhe the Link issues are. 
 Markdown result errorlength:        1 ../linter-result.out .
 Please run the Apply_Linter_check.sh in the tools folder to see where the Lint issues are.

File: Document/0x06c-Reverse-Engineering-and-Tampering.md (hunk 5 above)
Comment (cpholguera) [reply Document/0x06c-Reverse-Engineering-and-Tampering.md]: actually this is not really method hooking, but it is titled like that in master "Hooking Native Functions and Objective-C Methods". It is simply basic Cycript usage.

I'll merge it with Cycript's intro in "Dynamic Instrumentation -> Tooling". If someone wants to write about method hooking with Cycript we can open an issue (?)

File: Document/0x06b-Basic-Security-Testing.md (hunk 6 above)
Comment (cpholguera) [reply Document/0x06b-Basic-Security-Testing.md]: little img added :)

File: Document/0x06c-Reverse-Engineering-and-Tampering.md (hunk 7 above)
Comment (cpholguera) [reply Document/0x06c-Reverse-Engineering-and-Tampering.md]: I agree, this is the actual version in master. I've changed it in both 5c and 6c to:

```
Building a reverse engineering environment for free is possible. However, there are some commercial alternatives. The most commonly used are:
```

what do you think?

File: Document/0x06c-Reverse-Engineering-and-Tampering.md (hunk 8 above)
Comment (cpholguera) [reply Document/0x06c-Reverse-Engineering-and-Tampering.md]: actually to section "Repackaging and Re-Signing", done ;)

File: Document/0x06c-Reverse-Engineering-and-Tampering.md (hunk 9 above)
Comment (cpholguera) [reply Document/0x06c-Reverse-Engineering-and-Tampering.md]: done

File: Document/0x06c-Reverse-Engineering-and-Tampering.md (hunk 10 above)
Comment (cpholguera) [reply Document/0x06c-Reverse-Engineering-and-Tampering.md]: I guess this text is already very old haha
I fixed all of them :)

File: Document/0x06c-Reverse-Engineering-and-Tampering.md (hunk 11 above)
Comment (cpholguera) [reply Document/0x06c-Reverse-Engineering-and-Tampering.md]: I've enhanced the comments to explain that

File: Document/0x06c-Reverse-Engineering-and-Tampering.md (hunk 5 above)
Comment (cpholguera) [reply Document/0x06c-Reverse-Engineering-and-Tampering.md]: I've done that and referenced Cycript Manual - <http://www.cycript.org/manual/>

I've removed the printMethods because it actually does not exist (is not built-in) and replaced it with a built-in function: choose(). The rest is basically the same content as before, just a bit more organized now.

**Comment by OMTGreleaser** [comment]
Results for commit 7b021c6c2409988a899bfb27fb05d73790a26c51: Broken link result: [!] Error(s) found by the Links validator: 4 pages have dead links! Verbose output in /link-check-result.out .
 Please run the Apply_Link_Check.sh in the tools folder to see wherhe the Link issues are. 
 Markdown result errorlength:        2 ../linter-result.out .
 Please run the Apply_Linter_check.sh in the tools folder to see where the Lint issues are.

**Comment by OMTGreleaser** [comment]
Results for commit 4aa4013ebdc83b108992630cd2286dcc0ab1fc44: Broken link result: [!] Error(s) found by the Links validator: 4 pages have dead links! Verbose output in /link-check-result.out .
 Please run the Apply_Link_Check.sh in the tools folder to see wherhe the Link issues are. 
 Markdown result errorlength:        2 ../linter-result.out .
 Please run the Apply_Linter_check.sh in the tools folder to see where the Lint issues are.

**Comment by OMTGreleaser** [comment]
Results for commit e7025b18263c48e7ece26d8b43e8d9ec11e2fe3e: Broken link result: [!] Error(s) found by the Links validator: 5 pages have dead links! Verbose output in /link-check-result.out .
 Please run the Apply_Link_Check.sh in the tools folder to see wherhe the Link issues are. 
 Markdown result errorlength:        2 ../linter-result.out .
 Please run the Apply_Linter_check.sh in the tools folder to see where the Lint issues are.

File: Document/0x06c-Reverse-Engineering-and-Tampering.md (hunk 4 above)
Comment (commjoen) [reply Document/0x06c-Reverse-Engineering-and-Tampering.md]: I guess we should do this in another ticket some other time, but not now :).
